drivers/network/netio/netio.c

```diff
@@ -0,0 +1,1515 @@
+}
+
+static NTSTATUS WSKAPI
+WskControlClient(
+    _In_ PWSK_CLIENT Client,
+    _In_ ULONG ControlCode,
+    _In_ SIZE_T InputSize,
+    _In_reads_bytes_opt_(InputSize) PVOID InputBuffer,
+    _In_ SIZE_T OutputSize,
+    _Out_writes_bytes_opt_(OutputSize) PVOID OutputBuffer,
+    _Out_opt_ SIZE_T * OutputSizeReturned, _Inout_opt_ PIRP Irp)
+{
+    FUNCTION_TRACE;
+
+    UNIMPLEMENTED;
+    if (Irp != NULL)
+    {
+        Irp->IoStatus.Status = STATUS_NOT_IMPLEMENTED;
+        IoCompleteRequest(Irp, IO_NETWORK_INCREMENT);
+    }
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+static WSK_PROVIDER_DATAGRAM_DISPATCH UdpDispatch = {
+    .WskControlSocket = WskControlSocket,
+    .WskCloseSocket = WskCloseSocket,
+    .WskBind = WskBind,
+    .WskSendTo = WskSendTo,
+    .WskReceiveFrom = WskReceiveFrom,
+    .WskRelease = WskReleaseUdp,
+    .WskGetLocalAddress = WskGetLocalAddress,
+};
+
+/* Connection oriented routines (TCP/IP): */
+
+static NTSTATUS WSKAPI
+WskConnect(_In_ PWSK_SOCKET SocketParam, _In_ PSOCKADDR RemoteAddress, _Reserved_ ULONG Flags, _Inout_ PIRP Irp)
+{
+    PTDI_CONNECTION_INFORMATION TargetConnectionInfo, PeerAddrRet;
+    PIRP tdiIrp;
+    PWSK_SOCKET_INTERNAL Socket = CONTAINING_RECORD(SocketParam, WSK_SOCKET_INTERNAL, Socket);
+    NTSTATUS status, status2;
+    PNETIO_CONTEXT NetioContext;
+
+    FUNCTION_TRACE;
+
+    IoSetNextIrpStackLocation(Irp);
+
+    status = STATUS_INVALID_PARAMETER;
+    if (Socket->LocalAddressHandle == NULL)
+    {
+        DPRINT1("LocalAddressHandle is not set, need to bind your socket before connecting\n");
+        goto err_out;
+    }
+
+    status = STATUS_INSUFFICIENT_RESOURCES;
+    NetioContext = ExAllocatePoolUninitialized(NonPagedPool, sizeof(*NetioContext), TAG_NETIO);
+    if (NetioContext == NULL)
+    {
+        goto err_out;
+    }
+    NetioContext->socket = Socket;
+    NetioContext->UserIrp = Irp;
+
+    tdiIrp = NULL;
+
+    status = TdiAssociateAddressFile(Socket->LocalAddressHandle, Socket->ConnectionFile);
+    if (!NT_SUCCESS(status))
+    {
+        goto err_out_free_nc;
+    }
+    Socket->ConnectionFileAssociated = TRUE;
+
+    TargetConnectionInfo = TdiConnectionInfoFromSocketAddress(RemoteAddress);
+    if (TargetConnectionInfo == NULL)
+    {
+        goto err_out_free_nc_and_disassociate;
+    }
+    NetioContext->TargetConnectionInfo = TargetConnectionInfo;
+
+    PeerAddrRet = NULL;
+    TdiBuildNullConnectionInfo(&PeerAddrRet, TDI_ADDRESS_TYPE_IP);
+    if (PeerAddrRet == NULL)
+    {
+        goto err_out_free_nc_and_tci;
+    }
+    PeerAddrRet->OptionsLength = 0;
+    NetioContext->PeerAddrRet = PeerAddrRet;
+
+    IoMarkIrpPending(Irp);
+    SocketGet(Socket);
+
+    status = TdiConnect(&tdiIrp, Socket->ConnectionFile, TargetConnectionInfo, PeerAddrRet, NetioComplete, NetioContext);
+
+    /* If allocating tdiIrp fails we get here.
+     * Call the IoCompletion of the application's Irp so this Irp
+     * gets freed.
+     */
```

Comment:
This comment should probably be moved inside the `if (!NT_SUCCESS(status))` block, or down below just above the `err_out_free_nc_and_tci:` label.